In [ ]:
# import create_engine from sqlalchemy package to connect to MySQL 
from sqlalchemy import create_engine

In [ ]:
# import nbmerge package to merge notebooks
!pip install nbmerge

In [ ]:
# install ipython-sql library
!pip install ipython-sql

Load SQL extension

In [10]:
%load_ext sql

Connect to the AWS RDS bali-hotels database

In [ ]:
%sql mysql://USERNAME:PASSWORD@HOST/DATABASE

# SQL Analysis

## Part 1 Exploratory Queries

### 1. How many accomodation listings in Bali does Booking.com host?

In [39]:
%%sql
SELECT COUNT(*) AS hotel_count
FROM hotels;

 * mysql+mysqldb://admin:***@bali-hotels.cj8fr6zjuwf1.us-east-1.rds.amazonaws.com/bali_hotels?charset=utf8
1 rows affected.


[(755,)]

#### Insights
The web scrape of Bali hotels on Booking.com returned 1000 records. However, after data cleaning (e.g. removing duplicates) there are only 755 accomodation listings.

### 2. Find the lowest, highest, and average hotel rating

In [42]:
%%sql
SELECT 
    MIN(rating) AS lowest_rating, 
    MAX(rating) AS highest_rating, 
    ROUND(AVG(rating), 2) AS avg_rating
FROM hotels;

 * mysql+mysqldb://admin:***@bali-hotels.cj8fr6zjuwf1.us-east-1.rds.amazonaws.com/bali_hotels?charset=utf8
1 rows affected.


[(Decimal('0.0'), Decimal('10.0'), Decimal('8.53'))]

#### Insights
The lowest rating is 0, the highest is 10, and the average is 8.53. The lowest rating of 0 could be on account of the hotel being a new listing or other external factors. The maximum rating of 10 tells us that this is the highest possible rating a hotel can achieve. For further analysis, we could see how many and which hotels have a perfect rating.

### 3. Find the lowest, highest, and average hotel price

In [57]:
%%sql
SELECT 
    MIN(price) AS lowest_price, 
    MAX(price) AS highest_price, 
    ROUND(AVG(price), 2) AS avg_price
FROM hotels;

 * mysql+mysqldb://admin:***@bali-hotels.cj8fr6zjuwf1.us-east-1.rds.amazonaws.com/bali_hotels?charset=utf8
1 rows affected.


[(Decimal('3'), Decimal('1136'), Decimal('73.76'))]

#### Insights
The lowest price for a hotel in Bali is \\$3, the highest is \\$1,136, and the average is \\$73.76.

### 4. How many hotels are located in each district?

In [44]:
%%sql
SELECT 
    district_name, 
    COUNT(hotel_id) AS hotel_count_by_district
FROM hotels h
JOIN district d
    ON h.district = d.district_id
GROUP BY district_name
ORDER BY hotel_count_by_district DESC;

 * mysql+mysqldb://admin:***@bali-hotels.cj8fr6zjuwf1.us-east-1.rds.amazonaws.com/bali_hotels?charset=utf8
60 rows affected.


[('Ubud', 169),
 ('Seminyak', 85),
 ('Canggu', 66),
 ('Kuta', 51),
 ('Sanur', 40),
 ('Legian', 36),
 ('Nusa Penida', 35),
 ('Uluwatu', 26),
 ('Jimbaran', 25),
 ('Nusa Lembongan', 22),
 ('Nusa Dua', 20),
 ('Lovina', 16),
 ('Denpasar', 15),
 ('Amed', 15),
 ('Candidasa', 11),
 ('Tabanan', 10),
 ('Kubupenlokan', 8),
 ('Banyuwedang', 8),
 ('Pemuteran', 7),
 ('Tanah Lot', 6),
 ('Kerobokan', 6),
 ('Sidemen', 5),
 ('Keramas', 4),
 ('Tejakula', 4),
 ('Sukawati', 4),
 ('Munduk', 4),
 ('Tulamben', 4),
 ('Kintamani', 4),
 ('Pulukan', 3),
 ('Payangan', 3),
 ('Balian', 3),
 ('Ungasan', 3),
 ('Bedugul', 3),
 ('Airsatang', 2),
 ('Jatiluwih', 2),
 ('Tegalalang', 2),
 ('Kesiman', 2),
 ('Tampaksiring', 2),
 ('Karangasem', 2),
 ('Jasri', 2),
 ('Kubutambahan', 1),
 ('Klungkung', 1),
 ('Singaraja', 1),
 ('Ambengan', 1),
 ('Penginyahan', 1),
 ('Penginuman', 1),
 ('Pesanggaran', 1),
 ('Banjarangkan', 1),
 ('Bringkit', 1),
 ('Badung', 1),
 ('Buleleng', 1),
 ('Umeanyar', 1),
 ('Selemadeg', 1),
 ('Bangli', 1),
 ('Dalung', 1),
 ('Negara', 1),
 ('Seraya', 1),
 ('Tirtagangga', 1),
 ('Gianyar', 1),
 ('Sempidi', 1)]

#### Insights
The top 5 districts with the greatest number of hotels are Ubud, Seminyak, Canggu, Kuta, and Sanur.

### 5. What are all the unique subdistrict and district pairings and how many subdistricts does each district have?

In [56]:
%%sql
SELECT 
    subdistrict_id,
    district_id,
    subdistrict_name, 
    district_name,
    COUNT(subdistrict_id) OVER(
        PARTITION BY district_id
    ) AS subdistrict_count_by_district
FROM hotels h 
JOIN district d 
    ON h.district = d.district_id 
JOIN subdistrict s 
    ON h.subdistrict = s.subdistrict_id
GROUP BY subdistrict_id 
ORDER BY 
    subdistrict_count_by_district DESC,
    subdistrict_id;

 * mysql+mysqldb://admin:***@bali-hotels.cj8fr6zjuwf1.us-east-1.rds.amazonaws.com/bali_hotels?charset=utf8
62 rows affected.


[(2, 2, 'Pengosekan', 'Ubud', 10),
 (13, 2, 'Ubud City-Centre', 'Ubud', 10),
 (15, 2, 'Campuhan', 'Ubud', 10),
 (16, 2, 'Sambahan', 'Ubud', 10),
 (17, 2, 'Sayan', 'Ubud', 10),
 (20, 2, 'Taman', 'Ubud', 10),
 (30, 2, 'Andong', 'Ubud', 10),
 (48, 2, 'Penestanan', 'Ubud', 10),
 (57, 2, 'Kedewatan', 'Ubud', 10),
 (61, 2, 'Peliatan', 'Ubud', 10),
 (9, 12, 'Dyanapura', 'Seminyak', 9),
 (12, 12, 'Petitenget', 'Seminyak', 9),
 (14, 12, 'Sunset Road', 'Seminyak', 9),
 (29, 12, 'Batubelig', 'Seminyak', 9),
 (42, 12, 'Central Seminyak', 'Seminyak', 9),
 (43, 12, 'Drupadi', 'Seminyak', 9),
 (46, 12, 'Double Six', 'Seminyak', 9),
 (52, 12, 'Laksmana', 'Seminyak', 9),
 (53, 12, 'Nakula', 'Seminyak', 9),
 (3, 3, 'Berawa', 'Canggu', 8),
 (25, 3, 'Batu Bolong', 'Canggu', 8),
 (36, 3, 'Pererenan', 'Canggu', 8),
 (38, 3, 'Seseh', 'Canggu', 8),
 (39, 3, 'Umalas', 'Canggu', 8),
 (41, 3, 'Canggu Beach', 'Canggu', 8),
 (58, 3, 'Babakan', 'Canggu', 8),
 (62, 3, 'Padonan', 'Canggu', 8),
 (6, 7, 'Kampial', 'Nusa Dua', 6),
 (10, 7, 'Kutuh', 'Nusa Dua', 6),
 (23, 7, 'Tanjung Benoa', 'Nusa Dua', 6),
 (32, 7, 'BTDC', 'Nusa Dua', 6),
 (45, 7, 'Sawangan', 'Nusa Dua', 6),
 (55, 7, 'Mumbul', 'Nusa Dua', 6),
 (1, 1, 'Raya Kuta', 'Kuta', 5),
 (4, 1, 'Kartika Plaza', 'Kuta', 5),
 (21, 1, 'By Pass Ngurah Rai Kuta', 'Kuta', 5),
 (22, 1, 'Tuban', 'Kuta', 5),
 (27, 1, 'Downtown Kuta', 'Kuta', 5),
 (31, 34, 'Ungasan', 'Uluwatu', 5),
 (33, 34, 'Pecatu', 'Uluwatu', 5),
 (35, 34, 'Padang Padang', 'Uluwatu', 5),
 (49, 34, 'Bingin', 'Uluwatu', 5),
 (59, 34, 'Suluban', 'Uluwatu', 5),
 (5, 6, 'Padma', 'Legian', 4),
 (8, 10, 'Jimbaran Bay', 'Jimbaran', 4),
 (11, 6, 'Legian City-Centre', 'Legian', 4),
 (18, 10, 'Balangan', 'Jimbaran', 4),
 (19, 24, 'Diponegoro', 'Denpasar', 4),
 (34, 10, 'Bukit', 'Jimbaran', 4),
 (37, 24, 'Gatot Subroto', 'Denpasar', 4),
 (40, 6, 'Dewi Sri', 'Legian', 4),
 (44, 6, 'North Legian', 'Legian', 4),
 (47, 10, 'By Pass Ngurah Rai', 'Jimbaran', 4),
 (56, 24, 'Renon', 'Denpasar', 4),
 (60, 24, 'Teuku Umar', 'Denpasar', 4),
 (28, 33, 'Mushroom Bay', 'Nusa Lembongan', 3),
 (50, 33, 'Jungut Batu', 'Nusa Lembongan', 3),
 (51, 33, 'Nusa Ceningan', 'Nusa Lembongan', 3),
 (7, 8, 'Sanur Beach', 'Sanur', 1),
 (24, 28, 'Lovina Beach', 'Lovina', 1),
 (26, 30, 'Balian Beach', 'Selemadeg', 1),
 (54, 20, 'Medewi', 'Pulukan', 1)]

#### Insights
The top 3 districts with the highest number of subdistricts match the top 3 districts with the greatest number of hotels from the previous query.

## Part 2 Primary and Related Questions

### Business Problem
Booking.com wants to improve their customer experience by launching a new service to offer customers a helping hand in picking out the perfect hotel with our travel specialists. Instead of searching through filters and browsing through pages of hotels we want to test a new service that offers a personalised experience for each guest. Test this concept by curating a list of hotels for Customer X's trip.

### Customer X's Background, Preferences, and Requests
1. Tourist planning a solo trip to Bali
2. Overwhelmed by the choices on Booking.com
3. Prefers staying in popular areas
4. Willing to splurge and wants to treat themselves (first post-pandemic holiday!)

### 1. Identify the rating category for all hotels 

#### Business Justification
Assigning a descriptive word to the rating numbers will help us in this scenario where customers are more likely to provide their preferences and requests in terms of words rather than numbers.  

In [21]:
%%sql
CREATE OR REPLACE VIEW hotel_rating_category AS 
    SELECT 
        hotel_id,
        price,
        rating,
        CASE 
            WHEN rating > 8 THEN 'Wonderful'
            WHEN rating > 7 THEN 'Very Good'
            WHEN rating > 6 THEN 'Good'
            WHEN rating > 5 THEN 'Pleasant'
        ELSE 'Unsatisfactory' END AS rating_category
    FROM hotels h;

SELECT *
FROM hotel_rating_category;

 * mysql+mysqldb://admin:***@bali-hotels.cj8fr6zjuwf1.us-east-1.rds.amazonaws.com/bali_hotels?charset=utf8
755 rows affected.


[(1, Decimal('18'), Decimal('8.1'), 'Wonderful'),
 (2, Decimal('42'), Decimal('9.4'), 'Wonderful'),
 (3, Decimal('116'), Decimal('7.5'), 'Very Good'),
 (4, Decimal('17'), Decimal('9.4'), 'Wonderful'),
 (5, Decimal('81'), Decimal('8.6'), 'Wonderful'),
 (6, Decimal('92'), Decimal('9.3'), 'Wonderful'),
 (7, Decimal('121'), Decimal('9.2'), 'Wonderful'),
 (8, Decimal('50'), Decimal('8.9'), 'Wonderful'),
 (9, Decimal('104'), Decimal('9.4'), 'Wonderful'),
 (10, Decimal('20'), Decimal('7.6'), 'Very Good'),
 (11, Decimal('16'), Decimal('8.9'), 'Wonderful'),
 (12, Decimal('25'), Decimal('8.8'), 'Wonderful'),
 (13, Decimal('147'), Decimal('8.7'), 'Wonderful'),
 (14, Decimal('73'), Decimal('8.3'), 'Wonderful'),
 (15, Decimal('39'), Decimal('7.6'), 'Very Good'),
 (16, Decimal('40'), Decimal('9.0'), 'Wonderful'),
 (17, Decimal('158'), Decimal('8.3'), 'Wonderful'),
 (18, Decimal('109'), Decimal('9.0'), 'Wonderful'),
 (19, Decimal('107'), Decimal('8.8'), 'Wonderful'),
 (20, Decimal('126'), Decimal('9.2'), 'Wonderful'),
 (21, Decimal('19'), Decimal('8.5'), 'Wonderful'),
 (22, Decimal('33'), Decimal('9.2'), 'Wonderful'),
 (23, Decimal('36'), Decimal('7.9'), 'Very Good'),
 (24, Decimal('99'), Decimal('8.0'), 'Very Good'),
 (25, Decimal('39'), Decimal('7.9'), 'Very Good'),
 (26, Decimal('128'), Decimal('9.2'), 'Wonderful'),
 (27, Decimal('25'), Decimal('8.6'), 'Wonderful'),
 (28, Decimal('498'), Decimal('9.5'), 'Wonderful'),
 (29, Decimal('28'), Decimal('8.8'), 'Wonderful'),
 (30, Decimal('14'), Decimal('9.4'), 'Wonderful'),
 (31, Decimal('19'), Decimal('9.8'), 'Wonderful'),
 (32, Decimal('70'), Decimal('9.5'), 'Wonderful'),
 (33, Decimal('23'), Decimal('8.0'), 'Very Good'),
 (34, Decimal('132'), Decimal('9.4'), 'Wonderful'),
 (35, Decimal('37'), Decimal('8.8'), 'Wonderful'),
 (36, Decimal('43'), Decimal('9.7'), 'Wonderful'),
 (37, Decimal('42'), Decimal('8.4'), 'Wonderful'),
 (38, Decimal('21'), Decimal('9.3'), 'Wonderful'),
 (39, Decimal('270'), Decimal('9.2'), 'Wonderful'),
 (40, Decimal('34'), Decimal('8.6'), 'Wonderful'),
 (41, Decimal('12'), Decimal('8.1'), 'Wonderful'),
 (42, Decimal('108'), Decimal('9.2'), 'Wonderful'),
 (43, Decimal('80'), Decimal('9.3'), 'Wonderful'),
 (44, Decimal('53'), Decimal('8.8'), 'Wonderful'),
 (45, Decimal('48'), Decimal('9.1'), 'Wonderful'),
 (46, Decimal('61'), Decimal('9.1'), 'Wonderful'),
 (47, Decimal('19'), Decimal('9.5'), 'Wonderful'),
 (48, Decimal('118'), Decimal('9.4'), 'Wonderful'),
 (49, Decimal('65'), Decimal('9.0'), 'Wonderful'),
 (50, Decimal('71'), Decimal('9.0'), 'Wonderful'),
 (51, Decimal('76'), Decimal('9.1'), 'Wonderful'),
 (52, Decimal('94'), Decimal('8.7'), 'Wonderful'),
 (53, Decimal('44'), Decimal('8.2'), 'Wonderful'),
 (54, Decimal('49'), Decimal('10.0'), 'Wonderful'),
 (55, Decimal('15'), Decimal('7.6'), 'Very Good'),
 (56, Decimal('18'), Decimal('7.7'), 'Very Good'),
 (57, Decimal('12'), Decimal('8.9'), 'Wonderful'),
 (58, Decimal('40'), Decimal('9.3'), 'Wonderful'),
 (59, Decimal('9'), Decimal('8.3'), 'Wonderful'),
 (60, Decimal('24'), Decimal('9.0'), 'Wonderful'),
 (61, Decimal('24'), Decimal('8.4'), 'Wonderful'),
 (62, Decimal('19'), Decimal('8.5'), 'Wonderful'),
 (63, Decimal('18'), Decimal('9.4'), 'Wonderful'),
 (64, Decimal('79'), Decimal('9.2'), 'Wonderful'),
 (65, Decimal('8'), Decimal('9.0'), 'Wonderful'),
 (66, Decimal('70'), Decimal('8.4'), 'Wonderful'),
 (67, Decimal('62'), Decimal('7.6'), 'Very Good'),
 (68, Decimal('86'), Decimal('8.8'), 'Wonderful'),
 (69, Decimal('37'), Decimal('9.1'), 'Wonderful'),
 (70, Decimal('20'), Decimal('8.2'), 'Wonderful'),
 (71, Decimal('33'), Decimal('8.0'), 'Very Good'),
 (72, Decimal('11'), Decimal('7.4'), 'Very Good'),
 (73, Decimal('16'), Decimal('8.9'), 'Wonderful'),
 (74, Decimal('37'), Decimal('7.6'), 'Very Good'),
 (75, Decimal('16'), Decimal('8.5'), 'Wonderful'),
 (76, Decimal('46'), Decimal('9.4'), 'Wonderful'),
 (77, Decimal('133'), Decimal('7.8'), 'Very Good'),
 (78, Decimal('172'), Decimal('9.3'), 'Wonderful'),
 (79, Decimal('5

#### Recommendations
Since we have assigned the ratings with their respective category labels for each hotel, we can perform additional queries that group by the rating categories. Next, we should determine the average price for a hotel per rating category.

### 2. What is the average price for a hotel per rating category?

#### Business Justification
Understanding the average price of a hotel per rating category is a starting point to help us understand how much customers can expect to pay for a hotel based on the rating category.

In [27]:
%%sql
CREATE OR REPLACE VIEW avg_price_by_rating_category AS 
    SELECT 
        ROUND(AVG(price), 2) as avg_price, 
        rating_category
    FROM hotel_rating_category
    GROUP BY rating_category
    ORDER BY avg_price DESC;

SELECT * 
FROM avg_price_by_rating_category;

 * mysql+mysqldb://admin:***@bali-hotels.cj8fr6zjuwf1.us-east-1.rds.amazonaws.com/bali_hotels?charset=utf8
0 rows affected.
5 rows affected.


[(Decimal('80.19'), 'Wonderful'),
 (Decimal('48.73'), 'Unsatisfactory'),
 (Decimal('45.23'), 'Very Good'),
 (Decimal('27.88'), 'Good'),
 (Decimal('7.00'), 'Pleasant')]

#### Recommendations
1. There is approximately a 40% difference in average price between Wonderful and Very Good hotels. Surprisingly, the average price for Unsatisfactory hotels are higher than Very Good, Good, and Pleasant hotels. The lowest average price for a rating category is for Pleasant hotels. 

2. The business can consider performing a more in-depth sentiment analysis on hotel reviews to analyse other factors that impact average price. This can help the company factor in more values to provide greater accuracy in match results considering the variety of customers they will serve with this personalised service.

### 3a. Which hotels can we recommend to Customer X?

#### Business Justification
With an understanding of Customer X's background, preferences, and requests, the assigned travel specialist has determined that they should find hotels in the top 3 most popular districts where the rating category is Wonderful or Very Good. We can base the popularity of a district by the number of hotels located in the district. For Wonderful hotels filter hotels where the price is greater than or equal to the average price for a Wonderful hotel. For Very Good hotels, filter for hotels where the price is between the average price for these two categories. 

In [29]:
%%sql
SELECT
    DENSE_RANK() OVER (
        PARTITION BY district_name 
        ORDER BY rating DESC, h.price
    ) AS ranking,
    name,
    h.rating,
    h.price,
    rating_category,
    subdistrict_name AS subdistrict,
    district_name AS district,
    link
FROM hotels h 
JOIN district d 
    ON h.district = d.district_id
JOIN subdistrict s 
    ON h.subdistrict = s.subdistrict_id
JOIN hotel_rating_category hs
    ON h.hotel_id = hs.hotel_id 
WHERE district_name IN ('Ubud', 'Seminyak', 'Canggu')
    AND CASE 
            WHEN rating_category = 'Wonderful' THEN h.price >= 
                (SELECT avg_price FROM avg_price_by_rating_category WHERE rating_category = 'Wonderful')
            WHEN rating_category = 'Very Good' THEN h.price BETWEEN 
                (SELECT avg_price FROM avg_price_by_rating_category WHERE rating_category = 'Very Good')
                    AND (SELECT avg_price FROM avg_price_by_rating_category WHERE rating_category = 'Wonderful')
        ELSE NULL END;

 * mysql+mysqldb://admin:***@bali-hotels.cj8fr6zjuwf1.us-east-1.rds.amazonaws.com/bali_hotels?charset=utf8
95 rows affected.


[(1, 'Amar Boutique Hotel', Decimal('9.8'), Decimal('123'), 'Wonderful', 'Pererenan', 'Canggu', 'https://www.booking.com/hotel/id/amar-boutique.en-gb.html?label=gen173nr-1DCAEoggI46AdIM1gEaIkCiAEBmAExuAEHyAEM2AED6AEB-AECiAIBqAIDuALHr62SBsACAdICJDA2ZTc1ZjZlLTZkNTQtNGU1MC04ZTFkLTkxZWVhMDQwMmU5NtgCBOACAQ&sid=79ebc94bece7803777ea9717dcbea698&aid=304142&u'),
 (2, 'ZIN Canggu Resort & Villas', Decimal('9.4'), Decimal('144'), 'Wonderful', 'Batu Bolong', 'Canggu', 'https://www.booking.com/hotel/id/zin-canggu.en-gb.html?label=gen173nr-1DCAEoggI46AdIM1gEaIkCiAEBmAExuAEHyAEM2AED6AEB-AECiAIBqAIDuALHr62SBsACAdICJDA2ZTc1ZjZlLTZkNTQtNGU1MC04ZTFkLTkxZWVhMDQwMmU5NtgCBOACAQ&sid=79ebc94bece7803777ea9717dcbea698&aid=304142&ucfs'),
 (3, 'Villa Istana Putih by Nakula', Decimal('9.4'), Decimal('416'), 'Wonderful', 'Batu Bolong', 'Canggu', 'https://www.booking.com/hotel/id/villa-istana-putih.en-gb.html?label=gen173nr-1DCAEoggI46AdIM1gEaIkCiAEBmAExuAEHyAEM2AED6AEB-AECiAIBqAIDuALHr62SBsACAdICJDA2ZTc1ZjZlLTZkNTQtNGU1MC04ZTFkLTkxZWVhMDQwMmU5NtgCBOACAQ&sid=79ebc94bece7803777ea9717dcbea698&aid=304'),
 (4, 'Villa Del Mar Canggu', Decimal('9.3'), Decimal('124'), 'Wonderful', 'Batu Bolong', 'Canggu', 'https://www.booking.com/hotel/id/villa-del-mar-canggu.en-gb.html?label=gen173nr-1DCAEoggI46AdIM1gEaIkCiAEBmAExuAEHyAEM2AED6AEB-AECiAIBqAIDuALHr62SBsACAdICJDA2ZTc1ZjZlLTZkNTQtNGU1MC04ZTFkLTkxZWVhMDQwMmU5NtgCBOACAQ&sid=79ebc94bece7803777ea9717dcbea698&aid=3'),
 (5, 'Villa Sally', Decimal('9.3'), Decimal('292'), 'Wonderful', 'Berawa', 'Canggu', 'https://www.booking.com/hotel/id/villa-sally.en-gb.html?label=gen173nr-1DCAEoggI46AdIM1gEaIkCiAEBmAExuAEHyAEM2AED6AEB-AECiAIBqAIDuALHr62SBsACAdICJDA2ZTc1ZjZlLTZkNTQtNGU1MC04ZTFkLTkxZWVhMDQwMmU5NtgCBOACAQ&sid=79ebc94bece7803777ea9717dcbea698&aid=304142&ucf'),
 (6, 'Hotel Tugu Bali - CHSE Certified', Decimal('9.1'), Decimal('201'), 'Wonderful', 'Batu Bolong', 'Canggu', 'https://www.booking.com/hotel/id/tugu-bali.en-gb.html?label=gen173nr-1DCAEoggI46AdIM1gEaIkCiAEBmAExuAEHyAEM2AED6AEB-AECiAIBqAIDuALHr62SBsACAdICJDA2ZTc1ZjZlLTZkNTQtNGU1MC04ZTFkLTkxZWVhMDQwMmU5NtgCBOACAQ&sid=79ebc94bece7803777ea9717dcbea698&aid=304142&ucfs='),
 (7, 'COMO Uma Canggu - CHSE Certified', Decimal('9.1'), Decimal('241'), 'Wonderful', 'Batu Bolong', 'Canggu', 'https://www.booking.com/hotel/id/como-uma-canggu.en-gb.html?label=gen173nr-1DCAEoggI46AdIM1gEaIkCiAEBmAExuAEHyAEM2AED6AEB-AECiAIBqAIDuALHr62SBsACAdICJDA2ZTc1ZjZlLTZkNTQtNGU1MC04ZTFkLTkxZWVhMDQwMmU5NtgCBOACAQ&sid=79ebc94bece7803777ea9717dcbea698&aid=304142'),
 (8, 'Planta Luxury Boutique Resort', Decimal('9.0'), Decimal('127'), 'Wonderful', 'Batu Bolong', 'Canggu', 'https://www.booking.com/hotel/id/planta-villa.en-gb.html?label=gen173nr-1DCAEoggI46AdIM1gEaIkCiAEBmAExuAEHyAEM2AED6AEB-AECiAIBqAIDuALHr62SBsACAdICJDA2ZTc1ZjZlLTZkNTQtNGU1MC04ZTFkLTkxZWVhMDQwMmU5NtgCBOACAQ&sid=79ebc94bece7803777ea9717dcbea698&aid=304142&uc'),
 (9, 'Tapa Tepi Kali Echo Beach', Decimal('8.9'), Decimal('85'), 'Wonderful', 'Batu Bolong', 'Canggu', 'https://www.booking.com/hotel/id/tapa-tepi-kali.en-gb.html?label=gen173nr-1DCAEoggI46AdIM1gEaIkCiAEBmAExuAEHyAEM2AED6AEB-AECiAIBqAIDuALHr62SBsACAdICJDA2ZTc1ZjZlLTZkNTQtNGU1MC04ZTFkLTkxZWVhMDQwMmU5NtgCBOACAQ&sid=79ebc94bece7803777ea9717dcbea698&aid=304142&'),
 (10, 'THE HAVEN SUITES Bali Berawa', Decimal('8.9'), Decimal('86'), 'Wonderful', 'Berawa', 'Canggu', 'https://www.booking.com/hotel/id/the-haven-suite-bali-berawa.en-gb.html?label=gen173nr-1DCAEoggI46AdIM1gEaIkCiAEBmAExuAEHyAEM2AED6AEB-AECiAIBqAIDuALHr62SBsACAdICJDA2ZTc1ZjZlLTZkNTQtNGU1MC04ZTFkLTkxZWVhMDQwMmU5NtgCBOACAQ&sid=79ebc94bece7803777ea9717dcbea69'),
 (11, 'Aradhana Villas by Ekosistem', Decimal('8.8'), Decimal('120'), 'Wonderful', 'Berawa', 'Canggu', 'https://www.booking.com/hotel/id/aradhana-villas.en-gb.html?label=gen173nr-1DCAEoggI46AdIM1gEaIkCiAEBmAExuAEHyAEM2AED6AEB-AECiAIBqAIDuALHr62SBsACAdICJDA2ZTc1ZjZlLTZkNTQtNGU1MC04ZTFkLTkxZWVhMDQwMmU5NtgCBOACAQ&sid=79ebc94bece7803777ea9717dcbea698&aid=304142'),
 (12, 'Reg

#### Recommendations
Narrow down our options further for the customer.

### 3b. Which hotels should we recommend to Customer X?

#### Business Justification
We know that Customer X was interested in this new service because they were overwhelmed by the process of searching through pages of hotels. Expanding from the previous query, we should limit the list of hotel options to the top hotel from each district.

In [20]:
%%sql
WITH ranked_curated_hotels AS ( 
    SELECT
        DENSE_RANK() OVER (
            PARTITION BY district_name 
            ORDER BY rating DESC, h.price
        ) AS ranking,
        name,
        h.rating,
        h.price,
        rating_category,
        subdistrict_name AS subdistrict,
        district_name AS district,
        link
    FROM hotels h 
    JOIN district d 
        ON h.district = d.district_id
    JOIN subdistrict s 
        ON h.subdistrict = s.subdistrict_id
    JOIN hotel_rating_category hs
        ON h.hotel_id = hs.hotel_id 
    WHERE district_name IN ('Ubud', 'Seminyak', 'Canggu')
        AND CASE 
                WHEN rating_category = 'Wonderful' THEN h.price >= 
                    (SELECT avg_price FROM avg_price_by_rating_category WHERE rating_category = 'Wonderful')
                WHEN rating_category = 'Very Good' THEN h.price BETWEEN 
                    (SELECT avg_price FROM avg_price_by_rating_category WHERE rating_category = 'Very Good')
                        AND (SELECT avg_price FROM avg_price_by_rating_category WHERE rating_category = 'Wonderful')
            ELSE NULL END
)
SELECT *
FROM ranked_curated_hotels
WHERE ranking = 1
ORDER BY rating DESC;

 * mysql+mysqldb://admin:***@bali-hotels.cj8fr6zjuwf1.us-east-1.rds.amazonaws.com/bali_hotels?charset=utf8
3 rows affected.


[(1, 'Amar Boutique Hotel', Decimal('9.8'), Decimal('123'), 'Wonderful', 'Pererenan', 'Canggu', 'https://www.booking.com/hotel/id/amar-boutique.en-gb.html?label=gen173nr-1DCAEoggI46AdIM1gEaIkCiAEBmAExuAEHyAEM2AED6AEB-AECiAIBqAIDuALHr62SBsACAdICJDA2ZTc1ZjZlLTZkNTQtNGU1MC04ZTFkLTkxZWVhMDQwMmU5NtgCBOACAQ&sid=79ebc94bece7803777ea9717dcbea698&aid=304142&u'),
 (1, 'Villa Seminyak William', Decimal('9.7'), Decimal('530'), 'Wonderful', 'Petitenget', 'Seminyak', 'https://www.booking.com/hotel/id/villa-william-seminyak-bali.en-gb.html?label=gen173nr-1DCAEoggI46AdIM1gEaIkCiAEBmAExuAEHyAEM2AED6AEB-AECiAIBqAIDuALHr62SBsACAdICJDA2ZTc1ZjZlLTZkNTQtNGU1MC04ZTFkLTkxZWVhMDQwMmU5NtgCBOACAQ&sid=79ebc94bece7803777ea9717dcbea69'),
 (1, 'Villa Sindu Ubud', Decimal('9.6'), Decimal('97'), 'Wonderful', 'Sayan', 'Ubud', 'https://www.booking.com/hotel/id/villa-sindu-ubud.en-gb.html?label=gen173nr-1DCAEoggI46AdIM1gEaIkCiAEBmAExuAEHyAEM2AED6AEB-AECiAIBqAIDuALHr62SBsACAdICJDA2ZTc1ZjZlLTZkNTQtNGU1MC04ZTFkLTkxZWVhMDQwMmU5NtgCBOACAQ&sid=79ebc94bece7803777ea9717dcbea698&aid=30414')]

#### Recommendations
1. Now that we have the curated list of hotels for Customer X we should present these options to the customer. Once the customer approves of a hotel, we can guide them through the process of booking the hotel directly through the site. For the purpose of these tests, these queries can be modified for other customers.

2. The data science and analytics team should collaborate with the engineering team to develop an algorithm and program that can be used by the travel specialists to simplify this process. 

In [31]:
!nbmerge data_collection.ipynb sql_analysis.ipynb > presentation.ipynb